h/static/scripts/controllers/autosuggest-dropdown-controller.js

```diff
@@ -0,0 +1,431 @@
+let ACTIVE_ITEM_CLASS = '';
+let HEADER_CLASS = '';
+
+const ENTER = 13;
+const UP = 38;
+const DOWN = 40;
+
+/**
+ * Controller for adding autosuggest control to a piece of the page
+ */
+class AutosuggestDropdownController extends Controller {
+
+  /**
+   * @param {HTMLElement} inputElement that we are responding to in order to provide
+   *    suggestions. Note, we will add the suggestion container as a sibling to this
+   *    element.
+   * @param {Object} configOptions are used to set the initial set of items and the header
+   *    as well as providing the hooks for updates and callbacks
+   */
+  constructor(inputElement, configOptions) {
+
+    super(inputElement, configOptions);
+
+    if(!configOptions.renderListItem){
+      throw 'Missing renderListItem callback in AutosuggestDropdownController constructor';
+      return;
+    }
+
+    if(!configOptions.listReducer){
+      throw 'Missing listReducer function in AutosuggestDropdownController constructor';
+      return;
+    }
+
+    if(!configOptions.onSelect){
+      throw 'Missing onSelect callback in AutosuggestDropdownController constructor';
+      return;
+    }
+
+    // set up our element class attribute enum values
+    // Note, we currently are not doing anything with the default
+    // classes, but we have them if we wanted to give something for a default
+    // styling.
+    if(configOptions.classNames){
+        CONTAINER_CLASS = configOptions.classNames.container || 'autosuggest__container';
+        LIST_CLASS = configOptions.classNames.list || 'autosuggest__list';
+        ITEM_CLASS = configOptions.classNames.item || 'autosuggest__list-item';
+        ACTIVE_ITEM_CLASS = configOptions.classNames.activeItem || 'autosuggest__list-item--active';
+        HEADER_CLASS = configOptions.classNames.header || 'autosuggest__header';
+    }
+
+    // initial state values
+    this.setState({
+      visible: false,
+      header: configOptions.header,
+
+      // working list that are displayed to use
+      list: [],
+
+      // rootList is the original set that the reducer
+      // will receive to determine what should be shown
+      rootList: [],
+    });
+
+    // renaming simply to make it more obvious what
+    // the element is in other contexts the control
+    this._input = this.element;
+
+    // create the elements that make up the component
+    this._renderContentContainers();
+
+    this._addTopLevelEventListeners();
+
+    this._setList(configOptions.list);
+
+  }
+
+  // required to satisfy controller.js interface
+  update(){}
+
+  /**
+   * _setHeader - sets what would be the top bolded header text
+   *  to give context of what the suggestions are for
+   *
+   * @param  {string} header Text to place in header
+   */
+  _setHeader(header){
+    this.setState({
+      header
+    });
+    this._header.innerHTML = header;
+  }
+
+
+  /**
+   * _setList - update the current list
+   *
+   * @param  {Array} list The new list.
+   */
+  _setList(list){
```

Comment:
Just before I submitted the PR, I removed the public hooks to the select methods, like this one, that would be needed. I did that so to not conflate the needs of my next work with the needs of the refactor. So they will likely be coming back post refactor